src/librustc_mir/build/expr/into.rs

```diff
@@ -247,7 +247,11 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 } else {
                     let args: Vec<_> =
                         args.into_iter()
-                            .map(|arg| unpack!(block = this.as_local_operand(block, arg)))
+                            .map(|arg| {
```

Comment:
Actually, *can* we have a comment here that says that we must copy the argument to a temp, because functions are allowed to mutate arguments passed to them by-reference on the ABI level? And add a similar comment to `librustc/mir/mod.rs` so that there is no confusion?

Reply:
I'm not sure I like the resulting state from this... I think that `Call` should either take `Local` instead of `Operand`, *or* `trans` should generate the copy out of non-immediate constants.

Reply:
In fact, do we *really* want ABI concerns to shape MIR itself, as opposed to the ability to analyze it? I came back to this because I noticed that calls no longer had constants directly in them and thus copy propagation is incorrect *into* calls, and, without it, certain analyses would now need a bit more dataflow to even observe plain constants in function arguments.

Reply:
Talked this over with @arielb1 on IRC and I'm leaning towards "`Operand::Constant` is caller-owned, `Operand::Consume` is callee-owned", which would remove the temporaries for constant arguments and facilitate 

Furthermore, we can skip some copies (for `!Copy` arguments) during MIR construction and have MIR borrowck enforce the soundness of `Call`s (lock the return destination and move in all the arguments).